Release 0.12.0

zhilyaev opened the pull request:
not ready for merge yet!

zhilyaev commented:
also helm 3.6.0

r3nic1e commented on pkg/release/uninstall.go:
```diff
@@ -0,0 +1,81 @@
+	log.Info("🛥 Sync releases")
+	var fails []*Config
+
+	if feature.Parallel {
+		wg := parallel.NewWaitGroup()
+		log.Debug("🐞 Run in parallel mode")
+		wg.Add(len(releases))
+		for i := range releases {
+			go func(wg *parallel.WaitGroup, release *Config, fails *[]*Config) {
+				defer wg.Done()
+				release.UninstallWithFails(fails)
+			}(wg, releases[i], &fails)
+		}
+
+		err := wg.Wait()
+		if err != nil {
+			return err
+		}
+
+	} else {
+		for _, r := range releases {
+			r.UninstallWithFails(&fails)
+		}
+	}
```
I think it would be much easier to make WaitGroup with length of 1 instead of completely different code branch

r3nic1e commented on pkg/release/uninstall.go:
```diff
@@ -0,0 +1,81 @@
+		log.Debug("🐞 Run in parallel mode")
+		wg.Add(len(releases))
+		for i := range releases {
+			go func(wg *parallel.WaitGroup, release *Config, fails *[]*Config) {
```
Slices are already passed by reference (as they are references themselves), so getting its reference is pretty useless

r3nic1e commented on pkg/release/uninstall.go:
```diff
@@ -0,0 +1,81 @@
+)
+
+func (rel *Config) Uninstall() (*release.UninstallReleaseResponse, error) {
+	cfg, err := helper.ActionCfg(rel.Options.Namespace, helm.New())
```
TODO for future (probable should be an issue): there could be possible race with running helm commands with different namespaces. So we can make an RWMutex for it (maybe there are some fancy wrappers for this)

r3nic1e commented on pkg/action/build_test.go:
```diff
@@ -0,0 +1,122 @@
+	"github.com/helmwave/helmwave/tests"
+	"github.com/urfave/cli/v2"
+	"os"
+	"testing"
```
Maybe testify?

zhilyaev replied:
testify is pretty. But I have thoughts  that is complicated/overhead for our goals

r3nic1e commented on pkg/action/build_test.go:
```diff
@@ -0,0 +1,122 @@
+
+	b, _ := plan.NewBody(tests.Root + plan.Dir + plan.File)
+
+	if len(b.Releases) != 2 && b.Releases[0].Name != "redis-b" && b.Releases[1].Name != "memcached-b" {
+		t.Error("'redis-b' and 'memcached-b' not found")
+	}
```
checking length should have its own error

zhilyaev replied:
i am agree with you

r3nic1e commented on pkg/action/diff.go:
```diff
@@ -0,0 +1,65 @@
+		return err
+	}
+	if ok := plan1.IsManifestExist(); !ok {
+		return os.ErrNotExist
```
I think this should be wrapped into another error defining which plan caused this error

r3nic1e commented on pkg/action/diff.go:
```diff
@@ -0,0 +1,65 @@
+	}
+
+
+	_ = plan1.Diff(plan2)
+	return nil
```
why not to return `Diff()` error?

r3nic1e commented on pkg/action/diff.go:
```diff
@@ -0,0 +1,65 @@
+			},
+			&cli.StringFlag{
+				Name:        "plandir2",
+				Value:       ".helmwave/",
```
This is probably bad default value since it equals first plandir

zhilyaev replied:
No changes expected :)

r3nic1e commented on pkg/action/install.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+type Install struct {
+	dog      *kubedog.Config
```
🐶 should be kubedog to avoid confusion :)

r3nic1e commented on pkg/action/install.go:
```diff
@@ -0,0 +1,67 @@
+	dog      *kubedog.Config
+	plandir  string
+	parallel bool
+	kubedog  bool
```
kubedogEnabled?

zhilyaev replied:
kubedog means in install context as kubedogEnabled

r3nic1e commented on pkg/plan/build.go:
```diff
@@ -0,0 +1,203 @@
+	"strings"
+)
+
+// Build Принимает на вход ямл и опции по работе с ним.
```
translate :)

zhilyaev replied:
todo translate)

r3nic1e commented on pkg/plan/build.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+// checkTagInclusion checks where any of release tags are included in target tags.
+func checkTagInclusion(targetTags, releaseTags []string, matchAll bool) bool {
+	for _, t := range targetTags {
+		contains := helper.Contains(t, releaseTags)
+		if matchAll && !contains {
+			return false
+		}
+		if !matchAll && contains {
+			return true
+		}
+	}
+
+	return matchAll
+}
```
This function is duplicated in helpers

zhilyaev replied:
todo refactor

r3nic1e commented on pkg/plan/new.go:
```diff
@@ -0,0 +1,73 @@
+
+	plan := &Plan{
+		dir:       dir,
+		fullPath:  dir + File,
```
there is `path.Join` for that

zhilyaev replied:
yep.

zhilyaev commented:
1. Yaml enhancement
    - [x] removed `install` and `dry-ran` from option. Fix philosophy of  helmwave.
    - [x] added createNamespace

2. Parallel is default
    - Parallel is stable
    - You cant disable parallel

3.  helm-diff.
    - [x] diff with previous plan 
    - [x] diff plan1 and plan2
    - [ ] ? diff plan with live
    - [x] manifests are part of plan

4.  Values
    - [ ] get over http
    - [ ] ? get over git

5. completion
  - [x] zsh
  - [x] bash
  - [ ] powershell

6. Repositories.
  - [x] Parallel install. 
  - [x] added force. 
  - [ ] ?  helm oci

7. New Commands
  - [x] helmwave ls
  - [ ]  helmwave rollback
  - [ ]  helmwave down
  - [x] helmwave up
  - [x] helmwave validate
  - [x] helmwave diff
  - [x] helmwave completion



8. Kubedog
    - Comeback

9. new emoji

10. Three steps of workflow.
    0. Render helmwave.yml.tpl to helmwave.yml.
    1. build plan.
    2. Work with plan.
